wire/msgmixpairreq.go

```diff
@@ -0,0 +1,426 @@
+// Copyright (c) 2023 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package wire
+
+import (
+	"bytes"
+	"fmt"
+	"hash"
+	"io"
+
+	"github.com/decred/dcrd/chaincfg/chainhash"
+)
+
+const (
+	// MaxMixPairReqScriptClassLen is the maximum length allowable for a
+	// MsgMixPairReq script class.
+	MaxMixPairReqScriptClassLen = 32
+
+	// MaxMixPairReqUTXOs is the maximum number of unspent transaction
+	// outputs that may be contributed in a single mixpairreq message.
+	MaxMixPairReqUTXOs = 512 // XXX: PNOOMA
+
+	// MaxMixPairReqUTXOScriptLen is the maximum length allowed for the
+	// unhashed P2SH script of a UTXO ownership proof.
+	// XXX: might want to limit this to standard script sizes
+	MaxMixPairReqUTXOScriptLen = 16384 // txscript.MaxScriptSize
+
+	// MaxMixPairReqUTXOPubKeyLen is the maximum length allowed for the
+	// pubkey of a UTXO ownership proof.
+	MaxMixPairReqUTXOPubKeyLen = 33
+
+	// MaxMixPairReqUTXOSignatureLen is the maximum length allowed for the
+	// signature of a UTXO ownership proof.
+	MaxMixPairReqUTXOSignatureLen = 64
+)
+
+// MixPairReqUTXO describes an unspent transaction output to be spent in a
+// mix.  It includes a proof that the output is able to be spent, by
+// containing a signature and the necessary data needed to prove key
+// possession.
+type MixPairReqUTXO struct {
+	OutPoint  OutPoint
+	Script    []byte // Only used for P2SH
+	PubKey    []byte
+	Signature []byte
+}
+
+// MsgMixPairReq implements the Message interface and represents a mixing pair
+// request message.  It describes a type of coinjoin to be created, unmixed
+// data being contributed to the coinjoin, and proof of ability to sign the
+// resulting coinjoin.
+type MsgMixPairReq struct {
+	Signature    [64]byte
+	Identity     [33]byte
+	Expiry       uint32
+	MixAmount    int64
```

Comment:
Since this and `InputValue` below are signed integers, the decode and encode below should reject negative numbers.  

Alternatively they could be made unsigned, but since all of the other amounts are signed given they are often involved in signed arithmetic, it seems like leaving them signed and just rejecting negative values when encoding/decoding makes sense.

Reply:
Added checks during deserialization that all amounts are non-negative.